File: frontend/de/unika/ipd/grgen/ir/exprevals/ExternalType.java
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 4.4
 * Copyright (C) 2003-2015 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3 (see LICENSE.txt included in the packaging of this file)
 * www.grgen.net
 */

/**
 * @author Edgar Jakumeit
 */

package de.unika.ipd.grgen.ir.exprevals;

import java.util.Collection;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import de.unika.ipd.grgen.ir.*;

/**
 * IR class that represents external types.
 */
public class ExternalType extends InheritanceType {
	private List<ExternalFunctionMethod> externalFunctionMethods = new LinkedList<ExternalFunctionMethod>();
	private List<ExternalProcedureMethod> externalProcedureMethods = new LinkedList<ExternalProcedureMethod>();

	private Map<String, ExternalFunctionMethod> allExternalFunctionMethods = null;
	private Map<String, ExternalProcedureMethod> allExternalProcedureMethods = null;

	/**
	 * Make a new external type.
	 * @param ident The identifier that declares this type.
	 */
	public ExternalType(Ident ident) {
		super("node type", ident, 0, null);
	}

	public Collection<ExternalFunctionMethod> getExternalFunctionMethods() {
		return Collections.unmodifiableCollection(externalFunctionMethods);
	}

	public void addExternalFunctionMethod(ExternalFunctionMethod method) {
		externalFunctionMethods.add(method);
		method.setOwner(this);
	}

	public void addExternalProcedureMethod(ExternalProcedureMethod method) {
		externalProcedureMethods.add(method);
		method.setOwner(this);
	}

	public Collection<ExternalProcedureMethod> getExternalProcedureMethods() {
		return Collections.unmodifiableCollection(externalProcedureMethods);
	}

	private void addExternalFunctionMethods(ExternalType type) {
		for(ExternalFunctionMethod fm : type.getExternalFunctionMethods()) {
			// METHOD-TODO - what version is of relevance if alreay defined; override handling
			String functionName = fm.getIdent().toString();
			allExternalFunctionMethods.put(functionName, fm);
		}
	}

	private void addExternalProcedureMethods(ExternalType type) {
		for(ExternalProcedureMethod pm : type.getExternalProcedureMethods()) {
			// METHOD-TODO - what version is of relevance if alreay defined; override handling
			String procedureName = pm.getIdent().toString();
			allExternalProcedureMethods.put(procedureName, pm);
		}
	}

	public Collection<ExternalFunctionMethod> getAllExternalFunctionMethods() {
		if( allExternalFunctionMethods == null ) {
			allExternalFunctionMethods = new LinkedHashMap<String, ExternalFunctionMethod>();
			//overridingMembers = new LinkedHashMap<Entity, Entity>(); METHOD-TODO

			// add the members of the super types
			for(InheritanceType superType : getAllSuperTypes())
				addExternalFunctionMethods((ExternalType)superType);

			// add members of the current type
			addExternalFunctionMethods(this);
		}

		return allExternalFunctionMethods.values();
	}

	public Collection<ExternalProcedureMethod> getAllExternalProcedureMethods() {
		if( allExternalProcedureMethods == null ) {
			allExternalProcedureMethods = new LinkedHashMap<String, ExternalProcedureMethod>();
			//overridingMembers = new LinkedHashMap<Entity, Entity>(); METHOD-TODO

			// add the members of the super types
			for(InheritanceType superType : getAllSuperTypes())
				addExternalProcedureMethods((ExternalType)superType);

			// add members of the current type
			addExternalProcedureMethods(this);
		}

		return allExternalProcedureMethods.values();
	}

	/** Return a classification of a type for the IR. */
	public int classify() {
		return IS_EXTERNAL_TYPE;
	}
}

File: engine-net-2/src/libGr/Graph/ExternalType.cs
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 4.5
 * Copyright (C) 2003-2019 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3 (see LICENSE.txt included in the packaging of this file)
 * www.grgen.net
 */

// by Moritz Kroll, Edgar Jakumeit

using System;
using System.Collections.Generic;
using System.Reflection.Emit;
using System.Diagnostics;
using System.IO;
using System.Collections;

namespace de.unika.ipd.grGen.libGr
{
    /// A representation of an external type registered with GrGen.
    /// The bottom type of the external type hierarchy that is always available is type object.
    /// </summary>
    public abstract class ExternalType
    {
        public ExternalType(string name, Type type)
        {
            Name = name;
            Type = type;
        }

        public void InitDirectSupertypes(ExternalType[] directSuperTypes)
        {
            this.directSuperTypes = directSuperTypes;
        }

        /// <summary>
        /// The name of the type.
        /// </summary>
        public readonly String Name;

        /// <summary>
        /// The .NET type of the type.
        /// </summary>
        public readonly Type Type;

        /// <summary>
        /// Array containing all direct super types of this type, the readonly interface.
        /// </summary>
        public ExternalType[] DirectSuperTypes { get { return directSuperTypes; } }

        /// <summary>
        /// Array containing all direct super types of this type, the real array.
        /// </summary>
        private ExternalType[] directSuperTypes;

        /// <summary>
        /// Checks, whether this type is compatible to the given type, i.e. this type is the same type as the given type
        /// or it is a sub type of the given type.
        /// </summary>
        /// <param name="that">The type to be compared to.</param>
        /// <returns>True, if this type is compatible to the given type.</returns>
        public bool IsA(ExternalType that)
        {
            if(this == that)
                return true;
            for(int i = 0; i < directSuperTypes.Length; ++i)
                if(directSuperTypes[i].IsA(that))
                    return true;
            return false;
        }

        /// <summary>
        /// The number of function methods of this type.
        /// </summary>
        public abstract int NumFunctionMethods { get; }

        /// <summary>
        /// Enumerates all function methods of this type.
        /// </summary>
        public abstract IEnumerable<IFunctionDefinition> FunctionMethods { get; }

        /// <summary>
        /// Returns a function definition object for the given function method name.
        /// If this type does not have a function method with this name, null is returned.
        /// </summary>
        /// <param name="name">Name of the function method</param>
        /// <returns>The function definition of the function method matching the name, or null if there is no such</returns>
        public abstract IFunctionDefinition GetFunctionMethod(String name);

        /// <summary>
        /// The number of procedure methods of this type.
        /// </summary>
        public abstract int NumProcedureMethods { get; }

        /// <summary>
        /// Enumerates all procedure methods of this type.
        /// </summary>
        public abstract IEnumerable<IProcedureDefinition> ProcedureMethods { get; }

        /// <summary>
        /// Returns a procedure definition object for the given procedure method name.
        /// If this type does not have a procedure method with this name, null is returned.
        /// </summary>
        /// <param name="name">Name of the procedure method</param>
        /// <returns>The procedure definition of the procedure method matching the name, or null if there is no such</returns>
        public abstract IProcedureDefinition GetProcedureMethod(String name);

        /// <summary>
        /// Returns the name of the type.
        /// </summary>
        /// <returns>The name of the type.</returns>
        public override string ToString()
        {
            return Name;
        }
    }
}
